File: Week1/merge-two-sorted-lists/wangdong.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        
        if(l1 == NULL && l2 == NULL) {
            return NULL;
        }
        
        if(l1 == NULL) {
            return l2;
        }
        
        if(l2 == NULL) {
            return l1;
        }
        
        ListNode *l = NULL;
        
        if(l1->val < l2->val) {
            l = l1;
            l->next = mergeTwoLists(l1->next,l2);
        }
        
        else {
            l = l2;
            l->next = mergeTwoLists(l2->next,l1);
        }
        
        return l;
        
    }
};

File: Week3/Reverse-Linked-List-II/wangdong.java
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        
        ListNode before , after, start, end;
        before = after = start = end = null;
        
        ListNode helper = head;
        
        for(int i = 1; i < m; i++) {
            before = helper;
            helper = helper.next;
        }
        
        start = end = helper;
        for(int i = m; i <= n; i++) {
            after = helper.next;
            helper.next = end;
            end = helper;
            helper = after;
        }
        
        if(before == null) {
            head = end;
        }
        else {
            before.next = end;
        }
        
        start.next = after;
        
        return head;
    }
}
